File: 2018-minimum-space-wasted-from-packaging/minimum-space-wasted-from-packaging.dart
class Solution {
  int minWastedSpace(List<int> packages, List<List<int>> boxes) {
    const int mod = 1000000007;

    packages.sort();

    int n = packages.length;

    List<int> prefix = List.filled(n + 1, 0);
    for (int i = 0; i < n; i++) {
      prefix[i + 1] = prefix[i] + packages[i];
    }

    int ans = 1 << 62;

    for (var supplier in boxes) {
      supplier.sort();

      if (supplier.last < packages.last) continue;

      int waste = 0;
      int idx = 0;

      for (int b in supplier) {
        int next = upperBound(packages, idx, b);

        waste += (next - idx) * b - (prefix[next] - prefix[idx]);

        idx = next;
        if (idx == n) break;
      }

      if (waste < ans) ans = waste;
    }

    return ans == (1 << 62) ? -1 : ans % mod;
  }

  int upperBound(List<int> arr, int start, int target) {
    int l = start;
    int r = arr.length;

    while (l < r) {
      int mid = (l + r) >> 1;
      if (arr[mid] <= target) {
        l = mid + 1;
      } else {
        r = mid;
      }
    }

    return l;
  }
}

File: 2018-minimum-space-wasted-from-packaging/minimum-space-wasted-from-packaging.swift
class Solution {
    func minWastedSpace(_ packages: [Int], _ boxes: [[Int]]) -> Int {
        let MOD = 1_000_000_007

        let packages = packages.sorted()
        let n = packages.count

        var prefix = Array(repeating: Int64(0), count: n + 1)
        for i in 0..<n {
            prefix[i + 1] = prefix[i] + Int64(packages[i])
        }

        var ans = Int64.max

        for var supplier in boxes {
            supplier.sort()

            if supplier.last! < packages.last! {
                continue
            }

            var waste: Int64 = 0
            var idx = 0

            for b in supplier {
                let next = upperBound(packages, idx, b)

                waste += Int64(next - idx) * Int64(b) - (prefix[next] - prefix[idx])

                idx = next
                if idx == n {
                    break
                }
            }

            ans = min(ans, waste)
        }

        return ans == Int64.max ? -1 : Int(ans % Int64(MOD))
    }

    private func upperBound(_ arr: [Int], _ start: Int, _ target: Int) -> Int {
        var left = start
        var right = arr.count

        while left < right {
            let mid = (left + right) / 2
            if arr[mid] <= target {
                left = mid + 1
            } else {
                right = mid
            }
        }

        return left
    }
}